azure-iot-device/azure/iot/device/common/transport/mqtt/pipeline_stages.py

```diff
@@ -0,0 +1,645 @@
+        """
+        if self.previous:
+            self.previous._handle_pipeline_event(event)
+
+    def continue_op(self, op):
+        """
+        Helper function to continue a given operation by passing it to the next stage
+        in the pipeline.  If there is no next stage in the pipeline, this function
+        will fail the operation and call complete_op to return the failure back up the
+        pipeline.  If the operation is already in an error state, this function will
+        complete the operation in order to return that error to the caller.
+        """
+        if op.error:
+            self.complete_op(op)
+        elif not self.next:
+            logger.error("%s(%s): no next stage.  Returning error", self.name, op.name)
+            op.error = NotImplementedError()
+            self.complete_op(op)
+        else:
+            self.next.run_op(op)
+
+    def complete_op(self, op):
+        """
+        Helper function to complete an operation by calling it's callback function thus
+        returning the result of the operation back up the pipeline.
+        """
+        logger.info("{}({}): completing".format(self.name, op.name))
+        op.callback(op)
+
+    def continue_with_translated_op(self, original_op, translated_op):
+        """
+        Continue an operation using a translated operation that carries on the work
+        represented by the operation.  When the translated operation completes, its
+        callback will cause the original operation to complete with the same result.
+
+        This is useful when a generic operation (such as "enable feature" needs to be
+        converted into a more specific operation (such as "subscribe to mqtt topic").
+        In that case, a stage's _run_op function would call this function passing in
+        the original "enable feature" op and the translated "subscribe to mqtt topic"
+        op.  This function will pass the subscribe op down, and when the subscribe op
+        is completed, this function will cause the original op to complete.
+
+        This function is only really useful if there is no data returned in the
+        translated_op that might need to be copied back into the original_op before
+        completing it.  If data needs to be copied this way, some other method needs
+        to be used.  (or a "copy data back" function needs to be added to this function
+        as an optional parameter.)
+        """
+
+        def translated_op_complete(op):
+            logger.info(
+                "%s: %s complete-returning result in %s",
+                self.name,
+                translated_op.name,
+                original_op.name,
+            )
+            original_op.error = translated_op.error
+            self.complete_op(original_op)
+
+        translated_op.callback = translated_op_complete
+        self.continue_op(translated_op)
+
+    # TODO: connected and disconnect migth become pipeline events.  maybe.
```

Comment:
># TODO: connected and disconnect migth become pipeline events.  maybe. [](start = 4, length = 70)

At some point we should really figure out some consistent story for connect and disconnect. Connect and Disconnect have been the bane of my existence at the transport layer, they're so inconsistent and hard to deal with. Plus I think we have a whole bunch of undefined edge cases surrounding them.

But I think that's a thing to deal with later.